### ISSUE TITLE
[BUG] package-lock file does not lock packages

### ISSUE THREAD

**Issue by tejohnso** [npm/cli#4431]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

I have run `npm install` in a project with a package.json file indicating a dependency on A, being a git+ssh dependency with a `#semver` indicating minor and patch version number changes will match. package-lock.json is produced with a `resolved:` value pointing to a git+ssh specific commit sha for dependency A.

I update dependency A on github with an updated patch version.

I run `npm install` again, and even though the package-lock indicates a specific commit sha, the updated dependency A is retrieved and the commit sha is updated in package-lock.

### Expected Behavior

I expect that package-lock behaves as documented.

>package-lock.json is automatically generated for any operations where npm modifies either the node_modules tree, or package.json. It describes the exact tree that was generated, such that subsequent installs are able to generate identical trees, regardless of intermediate dependency updates.

Specifically, I expect subsequent installs to generate identical trees. This did not happen.

### Steps To Reproduce

1. On linux with npm `8.5.0`
2. Complete the steps as described in the clear and concise description above.
4. Notice package-lock is modified contrary to its documented purpose.

### Environment

- npm: 8.5.0
- Node.js: 17.4.0
- OS Name: Arch Linux
- System Model Name: ?
- npm config:
```ini
; node bin location = /usr/bin/node
; cwd = /home/user/development/projects/project
; HOME = /home/user
```

**Comment by ljharb** [comment]
It’s supposed to do this; you can run `npm ci` if you want to exactly match the lockfile.

**Comment by tejohnso** [comment]
Thanks for the tip. 

How is one supposed to glean that from [the documentation](https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json)?

> This file is intended to be committed into source repositories, and serves various purposes:
Describe a single representation of a dependency tree such that teammates, deployments, and continuous integration are guaranteed to install exactly the same dependencies.

Given the passage above, it seems this behaviour is In fact, *contrary* to the documentation.

**Comment by ljharb** [comment]
You’re right that the docs should be improved; the reality is that if you want guarantees, you can only install from a registry (which is the best practice anyways; git is for dev, not for consumption)

**Comment by tejohnso** [comment]
Yes, it seems like there's a lot of undocumented gotchas. And you seem to be aware of many :)

> if you want guarantees, you can only install from a registry 

 Interesting. Now you're introducing another apparent documentation deficit. Or is this documented somewhere?

> git is for dev, not for consumption

Again...where is this documented? [This](https://docs.npmjs.com/about-packages-and-modules) indicates that "Git URLs used for npm packages can be formatted in the following ways", with no mention at all that it is only meant for dev.

**Comment by ljharb** [comment]
OK, so: a) things don't have to be documented to be true; b) npm doesn't document community conventions or longstanding best practices; c) git is keyed by sha, not semver version, and the `#1.2.3` syntax just fetches a tag - since git tags are mutable, git itself fails to provide guarantees there.

**Comment by tejohnso** [comment]
a) ok but I'm not disputing that and in some cases things are documented in an *untrue* way as in [this](https://github.com/npm/cli/issues/4431#issuecomment-1043058528) comment

b) ok so aside from the contrary documentation case, you're saying the everything else here is "community convention"? That seems hard to believe but I'll take your word for it. But I'll also second your previous statement that "docs should be improved".

c) git tag mutability is not responsible for what is happening in this case, but regardless, entries in package-lock resolve to a specific sha even if package.json references a semver / tag. Which makes sense since you'd expect package-lock avoid mutability.

**Comment by ljharb** [comment]
I do agree that the lockfile should be pinning a specific sha. However, it's a fetch from a website - npm *always* has to fetch from it and install, and can't ever assume it's immutable, because not every git-based website will provide that guarantee.

**Comment by fritzy** [comment]
As pointed out, this is a documentation issue. We should improve the documentation for `package-lock`, `install`, and `ci`.

**Comment by connorshea** [comment]
Has this been improved? This seems like a major problem if it's still working this way with recent supply chain attacks.

**Comment by tejohnso** [comment]
The documentation for [npm install](https://docs.npmjs.com/cli/v11/commands/npm-install) is pretty clear now:

> If the lockfile's resolved versions satisfy the package.json ranges: npm uses the exact versions from package-lock.json to ensure reproducible builds across environments.

Note there are no caveats about where the repo is fetched from.

A test on npm version `10.9.2` confirms the behaviour matches the expectation now. After updating the dependency repo with a new patch version and re-running `npm install` in the test repo, the sha specified in the lockfile is no longer modified. The dependency stays locked to the previous sha.
